### PR TITLE
Fix applying operations in default.qubit.tf when using 9 or more wires

### PR DESCRIPTION
**Pull request by trbromley**
This PR resolves https://github.com/PennyLaneAI/pennylane/issues/835.

Currently, TensorFlow does not fully support slicing into tensors of rank 9 or above (see discussion in https://github.com/PennyLaneAI/pennylane/issues/835). This impacts `default.qubit.tf`, which uses slicing when applying certain gates such as CNOT through special tensor manipulation tricks. These gates are applied through special methods such as `DefaultQubit._apply_cnot()`.

The following special methods are available, with those having a red circle impacted by the above limitation:
- `_apply_x()` :green_circle: 
- `_apply_y()` :red_circle: 
- `_apply_z()` :red_circle: 
- `_apply_hadamard()` :red_circle: 
- `_apply_s()`, `_apply_t()` :red_circle: 
- `_apply_cnot()` :red_circle: 
- `_apply_swap()` :green_circle: 
- `_apply_cz()` :red_circle:
Most methods are affected. This is due to common use of the underlying `_apply_phase()` method.

Due to the above, this PR simply empties the `_apply_ops` dictionary (which contains these methods) in `DefaultQubitTF` if instantiated with more than 8 wires.

Additional tests are added. One test tries to apply the CNOT gate with nine wires and is marked as an `xfail`, allowing us to see if/when TF fixes this limitation.

### CODE REVIEW

File: tests/devices/test_default_qubit_tf.py [hunk 1]
```diff
@@ -320,6 +320,24 @@ def test_two_qubit_parameters(self, init_state, op, func, theta, tol):
         expected = func(theta) @ state
         assert np.allclose(res, expected, atol=tol, rtol=0)
 
+    def test_apply_ops_dict_above_8_wires(self):
+        """Test that the apply_ops dictionary is empty when more than 8 wires are used"""
+        dev = DefaultQubitTF(wires=9)
+        assert dev._apply_ops == {}
```
Comment (josh146) [inline tests/devices/test_default_qubit_tf.py:326]: Would it be good to also execute, and assert that the correct apply method was called? Execution itself could be mocked out to avoid using too many resources

File: tests/devices/test_default_qubit_tf.py [hunk 2]
```diff
@@ -320,6 +320,24 @@ def test_two_qubit_parameters(self, init_state, op, func, theta, tol):
...
         assert np.allclose(res, expected, atol=tol, rtol=0)
 
+    def test_apply_ops_dict_above_8_wires(self):
+        """Test that the apply_ops dictionary is empty when more than 8 wires are used"""
+        dev = DefaultQubitTF(wires=9)
+        assert dev._apply_ops == {}
+
+    @pytest.mark.xfail(
+        raises=tf.errors.UnimplementedError,
+        reason="Slicing is not supported for " "more than 8 wires",
```
Comment (josh146) [inline tests/devices/test_default_qubit_tf.py:330]:
```suggestion
        reason="Slicing is not supported for more than 8 wires",
```

Comment (josh146) [reply tests/devices/test_default_qubit_tf.py:330]: 💪

File: tests/devices/test_default_qubit_tf.py [hunk 3]
```diff
@@ -320,6 +320,24 @@ def test_two_qubit_parameters(self, init_state, op, func, theta, tol):
         expected = func(theta) @ state
         assert np.allclose(res, expected, atol=tol, rtol=0)
 
+    def test_apply_ops_dict_above_8_wires(self):
+        """Test that the apply_ops dictionary is empty when more than 8 wires are used"""
+        dev = DefaultQubitTF(wires=9)
+        assert dev._apply_ops == {}
+
+    @pytest.mark.xfail(
+        raises=tf.errors.UnimplementedError,
```
Comment (josh146) [inline tests/devices/test_default_qubit_tf.py:329]: You could set this `xfail` to be strict if you want. This will cause the test to fail if it unexpectedly passes (the default behaviour is not to do this).

**Review by josh146 — APPROVED** [review]
Thanks Tom for solving this so quickly!

**Comment by josh146** [comment]
Oh, don't forget to update the changelog!

File: tests/devices/test_default_qubit_tf.py (hunk 1 above)
Comment (trbromley) [reply tests/devices/test_default_qubit_tf.py:326]: Nice, have done that.

File: tests/devices/test_default_qubit_tf.py (hunk 3 above)
Comment (trbromley) [reply tests/devices/test_default_qubit_tf.py:329]: Nice, done!

File: tests/devices/test_default_qubit_tf.py (hunk 2 above)
Comment (trbromley) [reply tests/devices/test_default_qubit_tf.py:330]: Fixed

**Review by trbromley — COMMENTED** [review]
Thanks @josh146!

Good point about the changelog - I wasn't sure if it was needed since it is correcting a bug that is only in `master` - the special apply functionality has not been released yet.

File: pennylane/devices/default_qubit_tf.py [hunk 4]
```diff
@@ -25,7 +25,7 @@
         raise ImportError("default.qubit.tf device requires TensorFlow>=2.0")
 
 except ImportError as e:
-    raise ImportError("default.qubit.tf device requires TensorFlow>=2.0")
+    raise ImportError("default.qubit.tf device requires TensorFlow>=2.0") from e
```
Comment (antalszava) [inline pennylane/devices/default_qubit_tf.py:28]: Oh nice!

File: pennylane/devices/default_qubit_tf.py [hunk 5]
```diff
@@ -159,6 +159,11 @@ def __init__(self, wires, *, shots=1000, analytic=True):
         # prevent using special apply method for this gate due to slowdown in TF implementation
         del self._apply_ops["CZ"]
 
+        # prevent using special apply methods when using more than 8 wires due to limitations
+        # with TF slicing
+        if wires > 8:
+            self._apply_ops = {}
```
Comment (antalszava) [inline pennylane/devices/default_qubit_tf.py:165]: Would we want the unaffected (`PauliX` and `SWAP`) to remain in there?

File: tests/devices/test_default_qubit_tf.py [hunk 6]
```diff
@@ -320,6 +320,33 @@ def test_two_qubit_parameters(self, init_state, op, func, theta, tol):
...
+        dev.apply(queue)
+
+        spy.assert_called_once()
+
+    @pytest.mark.xfail(
+        raises=tf.errors.UnimplementedError,
+        reason="Slicing is not supported for more than 8 wires",
+        strict=True,
+    )
+    def test_apply_ops_above_8_wires_using_special(self):
```
Comment (antalszava) [inline tests/devices/test_default_qubit_tf.py:341]: It's a good idea to add this one :)

**Comment by antalszava** [comment]
Hi @trbromley, left a couple of comments but overall this would look good to me :slightly_smiling_face:

File: pennylane/devices/default_qubit_tf.py (hunk 5 above)
Comment (trbromley) [reply pennylane/devices/default_qubit_tf.py:170]: Potentially, but I think the gain is relatively small since they are less common gates, so maybe better to turn off for clarity.

File: tests/devices/test_default_qubit_tf.py [hunk 7]
```diff
@@ -320,6 +320,33 @@ def test_two_qubit_parameters(self, init_state, op, func, theta, tol):
         expected = func(theta) @ state
         assert np.allclose(res, expected, atol=tol, rtol=0)
 
+    def test_apply_ops_above_8_wires(self, mocker):
+        """Test that when 9 wires are used, the _apply_ops dictionary is empty and application of a
+        CNOT gate is performed using _apply_unitary_einsum"""
+        dev = DefaultQubitTF(wires=9)
```
[the remark this answers is not in the document]
Comment (trbromley) [reply tests/devices/test_default_qubit_tf.py:341]: That is an excellent point! We discussed offline and found that the difference is due to you having an earlier version of PyTorch (`v2.0.1`) whereas I was using `v2.3.0`.

I then decided to run the test suite in `v2.0.1` and realized that these special apply operations do not work with this earlier version - there is a failure when calculating the gradient. In fact, the special apply ops are incompatible with versions before 2.3.0 (quite recent!). So I thought the best remedy (done in 60de26d09fdb334db17276cf9e16cb077c5be1b5) is to only use the special apply ops for `v2.3.0` and above.

**Review by trbromley — COMMENTED** [review]
Thanks @antalszava, good spot regarding the version!

**Comment by trbromley** [comment]
Thanks @antalszava @josh146, requesting a re-review primarily because of the slightly nontrivial update in 60de26d09fdb334db17276cf9e16cb077c5be1b5 (putting an extra restriction on TF version)

**Review by antalszava — APPROVED** [review]
Looks good to me! :100: Nice solution :)

File: tests/devices/test_default_qubit_tf.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (josh146) [reply tests/devices/test_default_qubit_tf.py:341]: It might be good to consider deprecating TF < 2.1 at some point. I already dislike that our support for TF versions is all over the place depending on the module, and at the same time, v2.0 was notoriously buggy when it comes to complex numbers.

**Review by josh146 — APPROVED** [review]
